pymatgen/analysis/spectra/spectrum.py

```diff
@@ -0,0 +1,161 @@
+# coding: utf-8
+# Copyright (c) Pymatgen Development Team.
+# Distributed under the terms of the MIT License.
+
+from monty.json import MSONable
+import abc
+from scipy.interpolate import interp1d
+from pymatgen.symmetry.analyzer import SpacegroupAnalyzer
+from pymatgen import Structure
+import json
+import numpy as np
+
+"""
+This module defines classes to represent all spectra
+"""
+
+__author__ = "Chen Zheng"
+__copyright__ = "Copyright 2012, The Materials Project"
+__version__ = "2.0"
+__maintainer__ = "Chen Zheng"
+__email__ = "chz022@ucsd.edu"
+__date__ = "Aug 9, 2017"
+
+
+class Spectrum(abc.ABC):
+    """
+    Base class for Xanespectrum and Exafspectrum.
+    Not meant to be instantiated directly
+    """
+
+    @property
+    def spectrum_structure(self):
+        """
+        Returns the structure associated with the spectrum
+        """
+        return self.structure
+
+    @property
+    def spectrum_type(self):
+        """
+        Returns the spectrum type
+        """
+        return self.type
+
+    @property
+    def spectrum_data(self):
+        """
+        Returns the spectrum
+        """
+        return self.spectrum
+
+    @classmethod
+    @abc.abstractmethod
+    def from_file(cls, filename):
+        """
+        Initiate Spectrum from a filename
+        """
+        pass
+
+
+class Xanespectrum(Spectrum, MSONable):
+    def __init__(self, formula, absorption_specie, edge, structure, spectrum):
+        """
+        Create a XANES spectrum object
+        :param formula: The reduced formula of the structure associated with the spectrum
+        :param absorption_specie: absorption specie of the spectrum
+        :param edge: absorption spectrum edge
+        :param structure: materials structure associated with the spectrum, need to be a pymatgen structure object
+        :param spectrum: Spectrum data
+        """
+
+        self.formula = formula
+        self.absorption_specie = absorption_specie
+        self.edge = edge
+        self.structure = structure
+        self.spectrum = spectrum
+
+
+class XanesFEFF(Xanespectrum):
```

Comment:
Note that things like dir_name and FEFF input parameters generally should not be stored in the class. Since when do we do this for any pymatgen analysis object? If you wish, you can have the XANES object take in a `provenance` optional kwarg. That can store things like whether it is calculated by FEFF or whatever code with what parameters. Most of the methods within the XANES object should not operate on provenace data.

Just think about the Dos object. Do I classify it as a VASP DOS? I don't! Density of states is a generic concept that can be generated by VASP, ABINIT and many other codes. The DOS object needs to be flexible enough to deal with all of this. The computational code specific stuff should be part of the io parsing, but not pollute analysis. The way I would analyze a VASP generated DOS is the same as I would do for a ABINIT generated one.

Based on your strategy, you are going to have tends of XANES object each catering to one type of computational code and repeating the exact same methods. That is ridiculous.